### PR TITLE
HOTFIX- Sy-1054 gob to msgpack migration

### PR DESCRIPTION
**Pull request by emilbon99**
# Fix Pull Request Template

## Key Information

- **Linear Issue**: [SY-1054](https://linear.app/synnaxlabs/issue/SY-1054/gob-to-msgpack-migration)

## Description

Fixes an issue where we'd use gob to encode and decode values, which produces inconsistent results on program restart. Also implements a fallback decoding strategy to use gob to decode values from older versions of Synnax.

## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added sufficient regression tests to cover the changes to CI.
- [x] I have added relevant tests to cover the changes or exposing bugs.
- [x] I have verified code coverage targets are met.

## Additional Notes
- [ ] These changes deal with concurrency.
- [ ] These changes affect UI.

## Manual QA Additions

- [ ] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
  with necessary manual QA steps to test my change.

## Reviewer Checklist
- [ ] Sufficient test coverage of new additions.
- [ ] Verified all steps in readiness checklists.
- [ ] UI changes have been tested.
- [ ] style and formatting is consistent.
- [ ] Reviewed any relevant changes to concurrent code for safety. 
- [ ] Sufficient comments and clarity of code.

### CODE REVIEW

**Comment by Lham42** [comment]
label as hotfix

File: x/go/binary/codec.go [hunk 1]
```diff
@@ -247,3 +247,73 @@ func UnmarshalStringUint64(b []byte) (n uint64, err error) {
...
 	return n, err
 }
+
+// FallbackCodec wraps a set of Codecs. When the first Codec in the chain fails to
+// encode/decode a value, it falls back to the next Codec in the chain.
+type FallbackCodec struct {
+	// Codecs is the list of codecs to fallback on.
+	Codecs []Codec
+	// FallbackOnEncode indicates whether to fallback on encoding.
+	FallbackOnEncode bool
```
Comment (Lham42) [inline x/go/binary/codec.go:257]: can you think of a specific situation we would want to set this to true? For forward compatibility or just if the encoding fails?

File: aspen/internal/kv/operation.go [hunk 2]
```diff
@@ -27,7 +27,12 @@ const (
 	recovered
 )
 
-var codec = &binary.MsgPackCodec{}
+// codec used to be implemented by a gob codec, but we want to switch to msgpack.
+// Instead, we will use a fallback codec that tries msgpack to decode first, then gob.
+var codec = &binary.FallbackCodec{
+	Codecs:           []binary.Codec{&binary.MsgPackCodec{}, &binary.GobCodec{}},
+	FallbackOnDecode: true,
```
Comment (Lham42) [inline aspen/internal/kv/operation.go:34]: Can we explicitly set FallbackOnEncode even if its false

File: x/go/binary/codec.go [hunk 3]
```diff
@@ -247,3 +247,73 @@ func UnmarshalStringUint64(b []byte) (n uint64, err error) {
...
+		panic("[binary] - no codecs provided to FallbackCodec")
+	}
+	if !f.FallbackOnEncode {
+		return f.Codecs[0].Encode(ctx, value)
+	}
+	if f.FallbackOnEncode {
+		for _, c := range f.Codecs {
+			b, err = c.Encode(ctx, value)
+			if err == nil {
+				return
```
Comment (Lham42) [inline x/go/binary/codec.go:276]: what happens if there is an error through all the codecs?

File: x/go/binary/codec.go [hunk 4]
```diff
@@ -247,3 +247,73 @@ func UnmarshalStringUint64(b []byte) (n uint64, err error) {
...
+		panic("[binary] - no codecs provided to FallbackCodec")
+	}
+	if !f.FallbackOnDecode {
+		return f.Codecs[0].Decode(ctx, data, value)
+	}
+	if f.FallbackOnDecode {
+		for _, c := range f.Codecs {
+			err = c.Decode(ctx, data, value)
+			if err == nil {
+				return
```
Comment (Lham42) [inline x/go/binary/codec.go:295]: what happens if there is an error through all the codecs?

File: x/go/binary/codec.go [hunk 5]
```diff
@@ -247,3 +247,73 @@ func UnmarshalStringUint64(b []byte) (n uint64, err error) {
...
+		panic("[binary] - no codecs provided to FallbackCodec")
+	}
+	if !f.FallbackOnDecode {
+		return f.Codecs[0].DecodeStream(ctx, r, value)
+	}
+	if f.FallbackOnDecode {
+		for _, c := range f.Codecs {
+			err = c.DecodeStream(ctx, r, value)
+			if err == nil {
+				return
```
Comment (Lham42) [inline x/go/binary/codec.go:314]: what happens if there is an error through all the codecs?

File: x/go/binary/codec_test.go [hunk 6]
```diff
@@ -75,4 +76,28 @@ var _ = Describe("Codec", func() {
...
+			js := &binary.JSONCodec{}
+			gb := &binary.GobCodec{}
+			type abc struct {
+				Value int `json:"value"`
+			}
+			v := abc{Value: 12}
+			jsonB := MustSucceed(js.Encode(nil, v))
+			gobB := MustSucceed(gb.Encode(nil, v))
+			var res abc
+			fbc := &binary.FallbackCodec{
```
Comment (Lham42) [inline x/go/binary/codec_test.go:90]: would have a test case for the situation i ask about above

**Review by Lham42 — CHANGES_REQUESTED** [review]

File: x/go/binary/codec.go (hunk 3 above)
Comment (emilbon99) [reply x/go/binary/codec.go:276]: We return the error from the last codec

File: x/go/binary/codec_test.go (hunk 6 above)
Comment (emilbon99) [reply x/go/binary/codec_test.go:90]: Sure

File: x/go/binary/codec.go (hunk 1 above)
Comment (emilbon99) [reply x/go/binary/codec.go:257]: No, I can't

**Review by Lham42 — APPROVED** [review]

**Pull request merged**